### PR TITLE
Test/syncable ledger with masks

### PR DESCRIPTION
**Pull request by johnwu93**
Thank you for contributing to Coda! Please see `CONTRIBUTING.md` if you haven't
yet.

Tested the interaction between syncable ledger and merkle masks.

I just moved around modules outside the test module in `test_mask` to test reuse  the module `Make_maskable_and_mask_with_depth `

Checklist:

- [X] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Does this close issues? List them:

Closes #0000
Closes #0000

### CODE REVIEW

File: src/lib/syncable_ledger/test_mask.ml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+module Make (Input : sig
+  val depth : int
+
+  val mask_layers : int
+end) =
+struct
+  open Merkle_ledger_tests.Test_stubs
+
+  module Hash = struct
```
Comment (psteckler) [inline src/lib/syncable_ledger/test_mask.ml]: Except for `to_hash`, this module looks to be the same as the `Hash` in the test stubs, and `to_hash` doesn't appear to be used. Could you just use the `Hash` in the test stubs?

File: src/lib/syncable_ledger/test_mask.ml [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    module Base_db = Maskable_and_mask.Base_db
+    module Any_base = Maskable_and_mask.Any_base
+    module Base = Any_base.M
+    module Mask = Maskable_and_mask.Mask
+    module Maskable = Maskable_and_mask.Maskable
+    include Mask.Attached
+
+    (* Each account for a layer of a mask will all have the same balance.
+    Specifically, the top most layer will have a balance of `balance` and
+    every layer afterwards will have it's balance doubled *)
```
Comment (psteckler) [inline src/lib/syncable_ledger/test_mask.ml]: its

File: src/lib/syncable_ledger/test_mask.ml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    open Merkle_ledger_tests.Test_stubs
+    module Base_db = Maskable_and_mask.Base_db
+    module Any_base = Maskable_and_mask.Any_base
+    module Base = Any_base.M
+    module Mask = Maskable_and_mask.Mask
+    module Maskable = Maskable_and_mask.Maskable
+    include Mask.Attached
+
+    (* Each account for a layer of a mask will all have the same balance.
+    Specifically, the top most layer will have a balance of `balance` and
```
Comment (psteckler) [inline src/lib/syncable_ledger/test_mask.ml]: a bit confusing here, not sure how "top most" and "afterward" are related.

Does this mean: 

  All accounts in the first mask added, which is at layer 0, has the balance given by `balance`. For a mask at layer `n` with a balance `b`, a mask added at layer `n + 1`, accounts have balance `2 * b`.

File: src/lib/syncable_ledger/test_mask.ml [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    every layer afterwards will have it's balance doubled *)
+    let load_ledger num_accounts (balance : int) : t * 'a =
+      let db = Base_db.create () in
+      let maskable = Any_base.cast (module Base_db) db in
+      let keys = Key.gen_keys num_accounts in
+      (* All the parents will have certain values *)
+      let initial_balance_multiplier = (1 lsl Input.mask_layers) * balance in
```
Comment (psteckler) [inline src/lib/syncable_ledger/test_mask.ml]: For this test, I think readability is more important than efficiency. So probably better to write `Int.pow 2 Input.mask_layers` than to use a shift.

File: src/lib/syncable_ledger/test_mask.ml [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    include Mask.Attached
+
+    (* Each account for a layer of a mask will all have the same balance.
+    Specifically, the top most layer will have a balance of `balance` and
+    every layer afterwards will have it's balance doubled *)
+    let load_ledger num_accounts (balance : int) : t * 'a =
+      let db = Base_db.create () in
+      let maskable = Any_base.cast (module Base_db) db in
+      let keys = Key.gen_keys num_accounts in
+      (* All the parents will have certain values *)
```
Comment (psteckler) [inline src/lib/syncable_ledger/test_mask.ml]: I don't know what this comment means.

File: src/lib/syncable_ledger/test_mask.ml [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+          let action, _ =
+            Maskable.get_or_create_account_exn maskable key account
+          in
+          assert (action = `Added) ) ;
+      let mask = Mask.create () in
+      let attached_mask = Maskable.register_mask maskable mask in
+      (* On the mask, all the children will have different values *)
+      let rec construct_layered_masks iter child_balance parent_mask =
+        if iter = 0 then (
+          assert (balance = child_balance) ;
```
Comment (psteckler) [inline src/lib/syncable_ledger/test_mask.ml]: should use `Int.equal` rather than polymorphic equality

File: src/lib/syncable_ledger/test_mask.ml [hunk 7]
```diff
@@ -0,0 +1,108 @@
+open Core
+
+(* Input for testing the interaction between merkle *)
```
Comment (psteckler) [inline src/lib/syncable_ledger/test_mask.ml]: from the comment, not so clear what interactions are being tested

File: src/lib/syncable_ledger/test_mask.ml [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+          in
+          let action, _ =
+            Maskable.get_or_create_account_exn maskable key account
+          in
+          assert (action = `Added) ) ;
+      let mask = Mask.create () in
+      let attached_mask = Maskable.register_mask maskable mask in
+      (* On the mask, all the children will have different values *)
+      let rec construct_layered_masks iter child_balance parent_mask =
+        if iter = 0 then (
```
Comment (psteckler) [inline src/lib/syncable_ledger/test_mask.ml]: Int.equal

File: src/lib/syncable_ledger/test_mask.ml [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+    let to_hash = Fn.id
+
+    let empty_account = hash_account Account.empty
+  end
+
+  module Maskable_and_mask =
+    Merkle_ledger_tests.Test_mask.Make_maskable_and_mask_with_depth (Input)
+
+  module L = struct
```
Comment (psteckler) [inline src/lib/syncable_ledger/test_mask.ml]: `L`  could be `Ledger` or even `Layered_masks_ledger` to be more informative

File: src/lib/syncable_ledger/test_mask.ml [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+    type account = Account.t
+
+    type hash = Hash.t
+  end
+
+  module Root_hash = Hash
+
+  module SL =
+    Syncable_ledger.Make (L.Addr) (Account) (Hash) (Hash) (L)
```
Comment (psteckler) [inline src/lib/syncable_ledger/test_mask.ml]: the second `Hash` could be `Root_hash`, as explained above.

As is, it looks funny providing the same module twice.

File: src/lib/syncable_ledger/test_mask.ml [hunk 11]
```diff
@@ -0,0 +1,108 @@
+open Core
```
Comment (psteckler) [inline src/lib/syncable_ledger/test_mask.ml]: there's a file with the same name in `merkle_ledger_tests`, which could be confusing.

File: src/lib/merkle_ledger_tests/test_mask.ml [hunk 12]
```diff
@@ -2,687 +2,665 @@
 
 open Core
 open Test_stubs
-
```
[the remark this answers is not in the document]
Comment (johnwu93) [reply src/lib/merkle_ledger_tests/test_mask.ml]: Yes, I wanted to reused the definition of `Make_maskable_and_mask_with_depth`

File: src/lib/syncable_ledger/test_mask.ml (hunk 9 above)
Comment (johnwu93) [reply src/lib/syncable_ledger/test_mask.ml]: I can change that for all the test inputs. Yah, the module type for the test input are all abbreviations

**Pull request merged**